proposals/ytt/004-schema-validation/README.md

```diff
@@ -0,0 +1,1523 @@
+
+Note: if an author wishes to provide a custom violation message, they would use the corresponding built-in rule as a custom rule.
+
+#### @assert/validate-string-defaults
+
+> **TODO:**
+> - [ ] firm-up name of this annotation: is it a bug or a feature that it is type-specific?
+
+Defines what a valid value is for all descendants which hold a "string" value.
+
+```
+@assert/validate-string-defaults [rule0, rule1...] [,<named-rules>] [,when=] [,when_null_skip=]
+```
+
+See also: [Consideration: Setting validation defaults for strings in a schema overlay](#consideration-setting-validation-defaults-for-strings-in-a-schema-overlay).
+
+#### @ytt:assert.validate()
+
+Programmatically asserts that a given value is valid based on the set of supplied rules.
+
+```python
+assert.validate(key, value, rule0, rule1, ... [,<named-rules>] [,when=] [,when_null_skip=])
+```
+where:
+- `key` (`string`) — the name of the value being validated.
+- `value` (any) — the value being validated.
+
+This function is the programmatic equivalent of [@assert/validate](#assertvalidate).
+
+#### @ytt:assert.is_valid()
+
+Asserts any validations present on a Node (and descendents).
+
+```python
+assert.is_valid(node)
+```
+
+- if all validation rules are satisfied, nothing happens.
+- if any validation rule is violated, `assert.fail()`s with the corresponding validation message(s).
+
+
+#### Included Assertion Functions
+
+- the functions in this section are technically higher-order functions that _produce_ assertions:
+    ```python
+    def assert.min(x):
+      return lambda v: assert.fail("at least {}", x) if v >= x else None
+    end
+    ```
+    of which the return value _is_ an assertion.
+- violation messages should be consistent across these functions.
+- these functions should attempt to support as many types as possible:
+
+##### @ytt:assert.contains()
+
+##### @ytt:assert.ends_with()
+
+##### @ytt:assert.format()
+
+Supported formats:
+- "quantity" — https://kubernetes.io/docs/reference/glossary/?all=true#term-quantity
+- Additional keywords will be added as needs are clear. Likely candidates come from common domain-specific formats:
+[JSON Schema Validation: §7 Vocabularies for Semantic Content With "format"](https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.7).
+
+##### @ytt:assert.len()
+
+The length of the value is _exactly_ that of the given length.
+
+##### @ytt:assert.matches()
+
+##### @ytt:assert.max()
+
+The value is _at most_ (inclusive) of the given minimum.
+
+##### @ytt:assert.max_len()
+
+The length of the value is _at most_ (inclusive) that of the given minimum.
+
+
+##### @ytt:assert.min()
+
+The value is _at least_ (inclusive) of the given minimum.
+
+##### @ytt:assert.min_len()
+
+The length of the value is _at least_ (inclusive) that of the given minimum.
+
+##### @ytt:assert.not_null()
+
+When present, this validator is always checked first.
+
+##### @ytt:assert.one_not_null()
```

Comment:
@joaopapereira + @cari-lynn noted that name is a bit lossy — it's unclear where it's _at least_ one ... or _exactly_ one is not null.

Exploring other name possibilities...
```
#@assert/one_not_required
#@assert/precisely_one_not_null
```

Reply:
`#@assert/exactly_one_required`
`#@assert/exactly_one_not_null`
`#@assert/exactly_one_of`

I also think `#@assert/one_not_null` is reasonable in favor of simplicity.
Inspired by [mathematical wording](https://mathworld.wolfram.com/ExactlyOne.html#:~:text=%22Exactly%20one%22%20means%20%22one,(three%20and%20only%20three).) for "one and only one"

Reply:
Potential confusion between `one_of()` — which is the enum assertion and `exactly_one_of()` which would be the mutually exclusive/union assertion. 🤔

Reply:
Through usability testing (vmware-tanzu/carvel-ytt#707), we found that `assert.one_not_null()`, in context, was sufficiently intuitive.